feat: 🎸 add scheduled backup service with optional healthcheck

ghost opened the pull request:
libre database backup service running slightly modified automysqlbackup.
automysqlbackup modifed to enhance functionality inside of a docker
container and expose a healthcheck port.

- Backs up all/some databases/tables on schedule
- Near-production ready out of the box
- Supports one-off runs for easy script integration
- GPL-2.0 licensed work with fast upstream support
- Health-check possible on port 18080
- Fully configurable to Daily/Weekly/Monthly
- Ability to run pre- and post-backup scripts
- Error logging and possibility of deduplication
- Extensive configuration options

https://user-images.githubusercontent.com/94251352/144147679-dbc0ed0d-ce0b-4691-9963-19d48bba22ab.mp4

Commit:
feat: 🎸 add scheduled backup service with optional healthcheck

libre database backup service running slightly modified automysqlbackup.
automysqlbackup modifed to enhance functionality inside of a docker
container and expose a healthcheck port.

0xc0000007b commented on docker-compose.yml:
```diff
@@ -22,11 +24,29 @@ services:
+      PASSWORD: "example"
+      DBHOST: database
+      DBEXCLUDE: "performance_schema information_schema"
+      CRON_SCHEDULE: "0 22 * * 1-7"
```
At 22:00 on every day-of-week from Monday through Sunday.

rht replied:
It would be helpful to have a human-readable/understandable format of this cron schedule.

rht commented on docker-compose.yml:
```diff
@@ -79,5 +98,5 @@ networks:
   # docker-compose.yml can be configured to interact with the services in the
   # other, separate docker-compose.yml using external networks.
-  common:
+  common: {}
```
What is the reason for adding `{}` here?

rht commented on docker-compose.yml:
```diff
@@ -5,7 +5,9 @@
 version: '3'
-
+volumes:
+  mysql-data: {}
```
Why are we splitting away from `mountPoint`?

rht commented on docker-compose.dev.yml:
```diff
@@ -19,3 +28,11 @@ services:
+    environment:
+      USERNAME: wikiuser
+      PASSWORD: "example"
+      CRON_SCHEDULE: "0 */2 * * *"
```
It would be helpful to have a human-readable/understandable format of this cron schedule.

Commit:
docs: ✏️ add human-readable summary of cron schedules

0xc0000007b commented:
Script integration has begun here, to be transferred to Inblock org following script integration. https://github.com/0xc0000007b/docker-mediawikibackup/pull/1

Commit:
refactor: 💡 update pkc script with db backup and restore logic

Commit:
fix: 🐛 update pkc compose restore service handle

0xc0000007b commented:
Restart script for service-only changes:

```sh
docker-compose stop mediawikibackup && docker-compose rm mediawikibackup && docker-compose up -d
```

And here's a test plan:

```
pkc nuke
pkc setup
pkc start
[restart service]

[web] Login
[web] Upload image
pkc stop
[web] Reload page
pkc start
[web] Reload page

lazydocker
[view logs]
[view cron schedule]
[access shell]
ls -R

pkc backup
[web] remove image
pkc restore
[web] refresh page
```

FantasticoFox commented on docker-compose.yml:
```diff
@@ -5,7 +5,10 @@
+volumes:
+  mediawiki-image-data: {}
+  mysql-data: {}
+  mysql-backups-data: {}
```
Would be helpful to have it general 'database-backups-data' we use already mariadb...

0xc0000007b replied:
Making that change could decrease comprehension of the code if we add another database in the future. This is more consistent with our ENV vars as well. @rht what are your thoughts? I know you leaned towards using `database-` as well so LMK what you think.

Commit:
fix: 🐛 use correct container name for image restore

FantasticoFox commented on pkc:
```diff
@@ -296,16 +296,43 @@ run_setup() {
+create_new_backup() {
+    sqlfile=$(docker run --rm --volumes-from micro-pkc-mediawikibackup busybox \
+        find /backup/latest -name '*.sql.gz' | head -1)
+    imgfile=$(docker run --rm --volumes-from micro-pkc-mediawikibackup busybox \
```
Should be general for 'files' which means it could be pdf / word or whatever.

FantasticoFox commented on pkc:
```diff
@@ -296,16 +296,43 @@ run_setup() {
+            "$tmpdir/$(basename "$sqlfile")"
+        docker cp micro-pkc-mediawikibackup:"$imgfile" \
+            "$tmpdir/$(basename "$imgfile")"
+        tar -cf "backup-$(date +"%Y-%m-%d_%Hh%Mm.%A")".tar -C "$tmpdir" .
```
On iteration I would love to see <domain_id><$wgserver / DNS name>

0xc0000007b replied:
noted. doing so will benefit the ability to vertically scale the autobackup service

FantasticoFox commented on pkc:
```diff
@@ -296,16 +296,43 @@ run_setup() {
+    if [[ -n "$sqlfile" && -n "$imgfile" ]]; then
+        echo "Saving a copy of the most recent backup..."
+        tmpdir=$(mktemp -d || exit 1)
+        docker cp micro-pkc-mediawikibackup:"$sqlfile" \
```
Thats great, this allows us to add more stuff in our tarball.

FantasticoFox commented on pkc:
```diff
@@ -296,16 +296,43 @@ run_setup() {
-    docker exec -it micro-pkc-mediawiki /MediaWiki_Backup/toggle_read_only_off.sh /var/www/html
+create_new_backup() {
+    sqlfile=$(docker run --rm --volumes-from micro-pkc-mediawikibackup busybox \
+        find /backup/latest -name '*.sql.gz' | head -1)
```
We tested find output and the order was not predictable. We checked for alternatives and found 'sort' and 'ls' as a reliable alternative. @rht ht used ls -t (sort by time, newest first) as the previous solution.

0xc0000007b replied:
Nice catch and great advice. Will do.

0xc0000007b replied:
updated yesterday during pairing.

FantasticoFox commented on pkc:
```diff
@@ -296,16 +296,43 @@ run_setup() {
+        micro-pkc-mediawikibackup automysqlbackup
+}
+
+restore_from_latest_autobackup() {
```
Missing wiki locking and restarting of the containers. The containers hold state in memory cache which will lead to corrupted data if not flushed. It was in the file but you removed it.

0xc0000007b replied:
lock added yesterday during pairing session

Commit:
fix: 🐛 sort db and images by date descending

Commit:
refactor: 💡 auto-backups every 30 mins when enabled

Commit:
fix: 🐛 remove named volumes when pkc is nuked

rht committed:
chore: 🤖 Fix typo

rht committed:
chore: 🤖 Document docker-compose down -v flag

rht committed:
refactor: 💡 Use ls -t instead of find | sort -r

This is because ls -t explicitly sort by the file timestamp instead of
the file name. The file name will contain the domain ID, which will make
the sort -r output less predictable.

rht committed:
feat: 🎸 setup: Add flag to disable autobackup service

Commit:
chore: 🤖 mark volume read only

Commit:
WIP

Commit:
refactor: 💡 update based on feedback to use mod time

Commit:
refactor: 💡 add guards around backup and restore

Commit:
refactor: 💡 add comment on obscure code

Commit:
docs: ✏️ add todo for future refactor

rht committed:
fix: 🐛 Do not run backup process in background

rht committed:
feat: 🎸 Clean up temp dir during backup

rht committed:
fix: 🐛 restore: Exit early when no backup archive found

Commit:
fix: 🐛 suppress output to stdout while locking db

FantasticoFox committed:
fix: 🐛 update docker compose source to 1.29.2

Commit:
refactor: 💡 move ownership of lock and unlock to backup service

Commit:
chore: 🤖 inform user location of output tarball

rht committed:
fix: 🐛 restore: Error out & exit early if MW not yet installed
